compiler/rustc_parse/src/parser/item.rs

```diff
@@ -1216,6 +1216,20 @@ impl<'a> Parser<'a> {
 
     /// Parses an enum declaration.
     fn parse_item_enum(&mut self) -> PResult<'a, ItemInfo> {
+        if self.prev_token.is_keyword(kw::Enum) && self.token.is_keyword(kw::Struct) {
+            let mut err = self.struct_span_err(
+                self.prev_token.span.to(self.token.span),
+                "`enum` and `struct` are mutually exclusive",
+            );
+            err.span_suggestion(
+                self.prev_token.span.to(self.token.span),
+                "replace `enum struct` with `enum`",
+                "",
+                Applicability::MachineApplicable,
+            );
+            return Err(err);
```

Comment:
We should be able to recover from this if the next token is an identifier. If so, then just do `err.emit();` instead of `return Err(..);`